### PR TITLE
feat(ws): update list workspacekinds JSON payloads

### PR DESCRIPTION
**Pull request by yehudit1987**
Resolution details for #128.
We investigate this issue with referring to payload example provided in #127 .
Seems the example is not relevant with all it's fields to workspacekind resource. 
Currently the output for the sample yaml as resource input (jupyterlab_v1beta1_workspacekind.yaml) - 

```bash
curl -i localhost:4000/api/v1/workspacekinds
```

```json
{
   "data":[
      {
         "name":"jupyterlab",
         "spawner":{
            "display_name":"JupyterLab Notebook",
            "description":"A Workspace which runs JupyterLab in a Pod",
            "deprecated":false,
            "deprecation_message":"This WorkspaceKind will be removed on 20XX-XX-XX, please use another WorkspaceKind.",
            "hidden":false
         },
         "pod_template":{
            "pod_metadata":{
               "labels":{
                  "my-workspace-kind-label":"my-value"
               },
               "annotations":{
                  "my-workspace-kind-annotation":"my-value"
               }
            },
            "image_config":{
               "current":"jupyterlab_scipy_190",
               "desired":"jupyterlab_scipy_190",
               "redirect_chain":[
                  {
                     "source":"jupyterlab_scipy_180",
                     "target":"jupyterlab_scipy_190"
                  }
               ]
            },
            "pod_config":{
               "current":"tiny_cpu",
               "desired":"tiny_cpu",
               "redirect_chain":[
                  {
                     "source":"tiny_cpu",
                     "target":"tiny_cpu"
                  },
                  {
                     "source":"tiny_cpu",
                     "target":"small_cpu"
                  },
                  {
                     "source":"tiny_cpu",
                     "target":"big_gpu"
                  }
               ]
            }
         }
      }
   ]
}
```

### CODE REVIEW

**Review by ederign — CHANGES_REQUESTED** [review]
Hi guys, sorry for the lack of clear specification. And thanks for the first PR.

I've updated https://github.com/kubeflow/notebooks/issues/128 with the payload needed for this task. Could you please update it!

**Comment by yehudit1987** [comment]
New output: 

[ws_kind_output.json](https://github.com/user-attachments/files/18008724/ws_kind_output.json)

File: workspaces/backend/internal/models/helper_test.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * limitations under the License.
+ * /
+ */
+package models
+
+import (
+	"testing"
+)
+
+func TestGetOrDefaultWithRecovery(t *testing.T) {
```
Comment (ederign) [inline workspaces/backend/internal/models/helper_test.go:24]: we will not need this test class (in favor of explicit null check)

File: workspaces/backend/internal/models/helper.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ * /
+ */
+package models
+
+import (
+	"fmt"
+)
+
+// GetOrDefaultWithRecovery safely retrieves the value, returning the default value if a panic occurs or the value is nil.
```
Comment (ederign) [inline workspaces/backend/internal/models/helper.go:24]: We will not need this code in favor of an explicit null check.

File: workspaces/backend/internal/models/workspacekinds.go [hunk 3]
```diff
@@ -19,81 +19,114 @@
...
+	Logo               map[string]string `json:"logo"`
+	PodTemplate        PodTemplateModel  `json:"pod_template"`
 }
 
-type PodTemplateModel struct {
-	ImageConfig string        `json:"image_config"`
-	PodConfig   string        `json:"pod_config"`
-	Resources   ResourceModel `json:"resources"`
-}
+func BuildImageConfigValues(item *kubefloworgv1beta1.WorkspaceKind) []ImageConfigValue {
```
Comment (ederign) [inline workspaces/backend/internal/models/workspacekinds.go:38]: Do we need to export this method?

File: workspaces/backend/internal/models/workspacekinds.go [hunk 4]
```diff
@@ -19,81 +19,114 @@
...
+	return imageConfigValues
 }
 
-type SpawnerModel struct {
-	DisplayName        string `json:"display_name"`
-	Description        string `json:"description"`
-	Deprecated         bool   `json:"deprecated"`
-	DeprecationMessage string `json:"deprecation_message"`
-	Hidden             bool   `json:"hidden"`
+func BuildPodConfigValues(item *kubefloworgv1beta1.WorkspaceKind) []PodConfigValue {
```
Comment (ederign) [inline workspaces/backend/internal/models/workspacekinds.go:70]: Do we need to export this method?

**Review by ederign — CHANGES_REQUESTED** [review]
@yehudit1987 thank you so much for the PR and I've asked a few small changes. Sorry for the delay on this PR. I took some long time of PTO and I'm now back to work.

I'll make that the next iterations of review it will be much more quick.

**Comment by ederign** [comment]
Thank you @yehudit1987 ! looks good for me.

**Comment by ederign** [comment]
Sample output.
```yaml
// 20250115091628
// http://localhost:4000/api/v1/workspacekinds/jupyterlab

{
  "data": {
    "name": "jupyterlab",
    "display_name": "JupyterLab Notebook",
    "description": "A Workspace which runs JupyterLab in a Pod",
    "deprecated": false,
    "deprecation_message": "This WorkspaceKind will be removed on 20XX-XX-XX, please use another WorkspaceKind.",
    "hidden": false,
    "icon": {
      "url": "https://jupyter.org/assets/favicons/apple-touch-icon-152x152.png"
    },
    "logo": {
      "url": "https://upload.wikimedia.org/wikipedia/commons/3/38/Jupyter_logo.svg"
    },
    "pod_template": {
      "pod_metadata": {
        "labels": {
          "my-workspace-kind-label": "my-value"
        },
        "annotations": {
          "my-workspace-kind-annotation": "my-value"
        }
      },
      "volume_mounts": {
        "home": "/home/jovyan"
      },
      "options": {
        "image_config": {
          "default": "jupyterlab_scipy_190",
          "values": [
            {
              "id": "jupyterlab_scipy_180",
              "displayName": "jupyter-scipy:v1.8.0",
              "labels": {
                "python_version": "3.11"
              },
              "hidden": true,
              "redirect": {
                "to": "jupyterlab_scipy_190",
                "message": {
                  "text": "This update will change...",
                  "level": "Info"
                }
              }
            },
            {
              "id": "jupyterlab_scipy_190",
              "displayName": "jupyter-scipy:v1.9.0",
              "labels": {
                "python_version": "3.11"
              },
              "hidden": false
            }
          ]
        },
        "pod_config": {
          "default": "tiny_cpu",
          "values": [
            {
              "id": "tiny_cpu",
              "displayName": "Tiny CPU",
              "description": "Pod with 0.1 CPU, 128 Mb RAM",
              "labels": {
                "cpu": "100m",
                "memory": "128Mi"
              }
            },
            {
              "id": "small_cpu",
              "displayName": "Small CPU",
              "description": "Pod with 1 CPU, 2 GB RAM",
              "labels": {
                "cpu": "1000m",
                "memory": "2Gi"
              }
            },
            {
              "id": "big_gpu",
              "displayName": "Big GPU",
              "description": "Pod with 4 CPU, 16 GB RAM, and 1 GPU",
              "labels": {
                "cpu": "4000m",
                "gpu": "1",
                "memory": "16Gi"
              }
            }
          ]
        }
      }
    }
  }
}
```

**Review by ederign — APPROVED** [review]

**Comment by ederign** [comment]
@thesuperzapper this looks good for me and it is ready for your review.

**Comment by ederign** [comment]
@yehudit1987 can you fix the lint issue?

**Comment by thesuperzapper** [comment]
@yehudit1987 thanks for this PR, but please hold off for now, as @ederign and I are working on another one which changes the structure a bit.

**Comment by thesuperzapper** [comment]
Thanks for your effort, but this change has been superseded by https://github.com/kubeflow/notebooks/pull/192

As there was quite a bit of refactoring needed, especially in the tests.

/close
